### ISSUE TITLE
PredictLocation using deeploc error

### ISSUE THREAD

**Issue by SPD20** [SysBioChalmers/RAVEN#310]
### Description of the issue:

Hello, I am trying to incorporate location for my reactions. But I getting following error: Please find attached location file with my model.
[predict_location_juv.txt](https://github.com/SysBioChalmers/RAVEN/files/4977738/predict_location_juv.txt)
[juvenile_model_ori_gpr.xlsx](https://github.com/SysBioChalmers/RAVEN/files/4977749/juvenile_model_ori_gpr.xlsx)

Error using randsample (line 94)
W must contain non-negative values with at least one positive value.

Error in predictLocalization (line 386)
geneToMove=randsample(nGenes,1,true,max(GSS.scores(I,:),[],2)-GSS.scores(sub2ind(size(g2c),I,J))+0.1);

#### Reproducing this issue:

```matlab
[outModel, geneLocalization, transportStruct, score, removedRxns] = predictLocalization(newModel,GSS_juv,defaultCompartment,transportCost,maxTime)
```

#### System information
* Please report:
1. RAVEN version (stabile [release](https://github.com/SysBioChalmers/RAVEN/releases), `devel` branch?)
2. Operating system (Windows)

**I hereby confirm that I have:**
- [ X] Followed the [guidelines](https://github.com/SysBioChalmers/RAVEN/wiki/Installation) to install RAVEN.
- [ ] Checked that a similar issue does not [already exist](https://github.com/SysBioChalmers/RAVEN/issues?utf8=%E2%9C%93&q=is%3Aissue)
- [X ] If suitable, needed, asked first in the [Gitter chat room](https://gitter.im/SysBioChalmers/RAVEN) about the issue

**Comment by edkerk** [comment]
@SPD20 Could you please attach the model as SBML or MAT file, so we can replicate what you're trying to do?

**Comment by SPD20** [comment]
Hi,
Can I email the mat format model to you? Here its not letting me upload mat file saying that ..this kind of files are not supported.

Thanks

**Comment by edkerk** [comment]
You could ZIP the file first, then you should be able to attach it. Otherwise, you can email me at eduardk@chalmers.se.

**Comment by edkerk** [comment]
@SPD20 I've received your model file, but I'm not able to replicate what you encounter. What are the exact commands that you run before `predictLocalization`?

**Comment by SPD20** [comment]
Hello,

Thanks for your reply and please accept my apology for not sending enough information.

This time I loaded my model file in xml format using ' model=importModel' function although it was loaded but with warnings

WARNING: The composition for the following metabolites could not be parsed:
	(5')ppPur-mRNA[Cytoplasm]
	2-Methylthio-N6-L-threonylcarbamoyladenine-in-tRNA[Cytoplasm]
	5'-(N7-Methyl-5'-triphosphoguanosine)-(2'-O-methyl-purine-ribonucleotide)-(2'-O-methyl-ribonucleotide)-(mRNA)[Cytoplasm]
	5'-Phospho-(mRNA)[Cytoplasm]
	5'-Triphospho-(mRNA)[Cytoplasm]
	.....and 17 more
  
I changed my parsescore code as per your suggestion..but I was still getting following error

Error using cell/unique (line 85)
Cell array input must be a cell array of character vectors.

Error in parseScores_me (line 111)
[~, J, K]=unique(GSS.genes);

my location file(generated using deeploc tool) was comma separated here.

Then I changed my input file to tab separated and ran it with new parsescore code and got following error

Subscripted assignment dimension mismatch.

Error in parseScores_me (line 106)
        GSS.scores(row,:)=str2double(tline(4:end));

Finally I ran this tab separated file with old parse score code(attached) and got some results for GSS. 

GSS_juv = 

  struct with fields:

    compartments: {1×11 cell}
          scores: [1123×11 double]
           genes: {1123×1 cell}

My predictlocation code is still giving the error ...
plotResults = false
maxTime = 15
transportCost = 0.5
defaultCompartment = 'Cytoplasm'
[outModel, geneLocalization, transportStruct, score, removedRxns] = predictLocalization_new(model,GSS_juv,defaultCompartment,transportCost,maxTime)

Error using randsample (line 94)
W must contain non-negative values with at least one positive value.

Error in predictLocalization_new (line 402)
    toComp=randsample(nComps,1,true,GSS.scores(geneToMove,:)+0.2);

I will try to rectify problems of my model file...just in case they are behind this issue...

Thanks
[predict_location_juv2.txt](https://github.com/SysBioChalmers/RAVEN/files/5076573/predict_location_juv2.txt)

[parseScores_old.txt](https://github.com/SysBioChalmers/RAVEN/files/5076647/parseScores_old.txt)

**Comment by edkerk** [comment]
Right, it seems like there is some issues with different sources of DeepLoc (online vs. offline) giving slightly different output (there is difference in comma vs tab-separated, naming and formatting of columns). We're currently figuring out what the correct formats are, this should then fix the problems you encountered.

As a work around for you for now, from your DeepLoc file `parseScores` should ignore the third column, which specifies whether the protein is membrane or soluble. To do this, change parseScores as:

line 93: `GSS.compartments=GSS.compartments(4:end);`
line 106: `GSS.scores(row,:)=str2double(tline(4:end));` 

As was also done here: https://github.com/SysBioChalmers/RAVEN/commit/ff8629d55537e345a8212d2c34752c8e117970fe#diff-479365171760a0ecbfc3b9e4f4bbbc94. 

Apologies for this, it all seems to stem from different output from DeepLoc.

**Comment by SPD20** [comment]
Dear Eduard,

Thank you very much for  this workaround. I will implement it and will let you know if it worked.

Thanks,
Sonal

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by SPD20** [comment]
Dear Eduard,

Your workaround has worked and my GSS is giving some output. Is there anything I need to change in PredictLocalization script as well? As I am getting the following error.

[outModel, geneLocalization, transportStruct, score, removedRxns] = predictLocalization_edit(newModel,GSS_juv,defaultCompartment,transportCost,maxTime)

Error using vertcat
Dimensions of matrices being concatenated are not consistent.

Error in predictLocalization_edit (line 603)
outModel.compNames=[outModel.compNames;GSS.compartments(2:end)];

Can you please send me deeploc output used/preferred by Predictlocalization script, I will edit my results as per it?

Thanks,

**Comment by edkerk** [comment]
@eiden309 You have been running predictLocalization, did you run into these issues? And how does the output look like that you are feeding into the function?

**Comment by eiden309** [comment]
@edkerk yes I ran into the same issue previously, there is a minor bug in the function but it can be rectified by transposing `GSS.compartments(2:end)` in predictLocalization (line 611) i.e. adding `'` at the end of `GSS.compartments(2:end)`:

current: `outModel.compNames=[outModel.compNames;GSS.compartments(2:end)];`
new: `outModel.compNames=[outModel.compNames;GSS.compartments(2:end)'];`

I think that the output from `parseScores` can be used directly as the input for predictLocalization. @SPD20 FYI, if you still need the DeepLoc output file used for parseScores please do let me know. Hope this helps!
